# <img border="0" src="https://www.svgrepo.com/show/1025/task.svg" width="40" height="40"> Projects

***

<br/>

Since each of the 4 group themes have completely different softwares, please follow in instructions specific to each one.

Please follow the [pre-course instructions](precourse.md) in order to install all necessary software and packages.

The proposed educational plan for single cell RNA-seq data analysis project will be done using the Project-based learning (a.k.a. PBL) approach. This way, rather than having lectures on a topic and defined set of instructions and pre-given steps to follow (which focus on passive individual learning), you are required to learn the topic in a more **active** and **dynamic** way in order to solve the data analysis tasks as a group. Therefore, each group is provided with:

1. The <img border="0" src="https://static.thenounproject.com/png/67360-200.png" width="10" height="10">[report file](single_cell/code/project_report.html) containing a list of instructions and questions giving the overall direction and order of the steps needed to be done
2. A <img border="0" src="https://d1nhio0ox7pgb.cloudfront.net/_img/o_collection_png/green_dark_grey/512x512/plain/dictionary.png" width="10" height="10">[glossary of terms](single_cell/glossary/glossary_of_terms_single_cell.html) which contains detailed information about many of the steps and provides further references on the topics

In this manner, each group is required to gather information from the **glossary** to perform single cell data analysis by filling the report file (using this <img border="0" src="https://static.thenounproject.com/png/67360-200.png" width="10" height="10">[project_report.Rmd](single_cell/code/project_report.Rmd) file), by:

1. Reading the instructions for a certain task
2. Searching the **glossary** for guidance
3. Discussing with your colleges which step needs to be done or included
4. Dividing tasks among your group colleagues (each one can test parts of the code and then send the result to the one responsible for the report)
5. Replacing the instructions with a text explaining the rationale for this step and
6. Adding and running code from that step

You are encouraged to use any other additional website / article in your report. That also includes addition of other previous experiences and/or additional code steps from other sources, as long as the rationale for the inclusion is also discussed in the report text. You can also add any other plots and visualisations in your report to illustrate your results.

In the end, each group is expected to have 1 report file with all analysis steps and the rationale for the following projects:


# <img border="0" src="/single-cell_sib_scilifelab_2021/logos/single_cell.png" width="40" height="40"> Spatial Transcriptomics
***

<br/>

<details>
<summary>
<b>PROJECT 1: Blocking myeloid development during colitis</b>
</summary>

<br/>

**Background:** Ulcerative colitis (UC) is an inflammatory bowel disease (IBD) driven mainly by colonic innate inflammatory cells such as macrophages, monocytes and neutrophils ( [Czarnewski et al 2019](https://www.nature.com/articles/s41467-019-10769-x), [Skatteborg et al 2020](https://academic.oup.com/ecco-jcc/advance-article-abstract/doi/10.1093/ecco-jcc/jjaa121/5859161?redirectedFrom=fulltext)). A recent study showed that patients that present higher neutrophilic inflammatory signature (known as UC1) become refractory to both anti-TNF and anti-a4b7 integrin therapy ( [Czarnewski et al 2019](https://www.nature.com/articles/s41467-019-10769-x), [Skatteborg et al 2020](https://academic.oup.com/ecco-jcc/advance-article-abstract/doi/10.1093/ecco-jcc/jjaa121/5859161?redirectedFrom=fulltext)), which leads to surgical intervention for removal of the colon. Neutrophils inflammatory cells are short lived and originate from the common myeloid progenitor (CMP) in the bone marrow and requires constant replenishment in order to sustain elevated cell number in the colon. Herein, our main goal is to identify potential gene candidates that can block either pathways of neutrophil differentiation in the bone marrow.

**Main research question:** Which genes specifically drive the differentiation of Neutrophils.

**Importance:** Identifying such genes will allow us to: 1) perform experiment in Tamoxifen-transgenic mice where those cells can be depleted during the course of colitis. 2) find potential drugs that can inhibit those genes/pathways in order to block myeloid cell differentiation during colitis in mice (with priority to already approved drugs).

**Analysis report from group 1:**
1. [project_report_colitis](single_cell/group_reports/project_report_colitis.html) ([.Rmd](single_cell/group_reports/project_report_colitis.Rmd))
2. [Presentation](single_cell/group_reports/project_presentation_colitis.pdf)

</details>

<br/>

<details>
<summary>
<b>PROJECT 2: Identifying cell and gene candidates in severe COVID-19 patients</b>
</summary>

<br/>

**Background:** COVID-19 is an infectious disease driven by the virus SARS-CoV-2, which primarily infects lung epithelial cells. However, elderly patients usually develop severe lung inflammation and lung disfunction, ultimately leading to respiratory failure ([Guan et al 2020](https://www.nejm.org/doi/full/10.1056/nejmoa2002032)). The onset of the disease is characterised by a cytokine storm comprising several inflammatory mediators ([Pedersen et al 2020](https://www.jci.org/articles/view/137647)), specially in severe cases of the disease. Many cell types orchestrate the immune response to the virus, but their relative contribution at the single-cell resolution is still unclear. Herein, our main goal is to identify which cell types and gene pathways are altered in the blood of patients with severe COVID-19.

**Main research question:** Which cell types and genes are altered when comparing blood immune cells from healthy versus COVID-19 patients.

**Importance:** Identifying such genes will allow us to: 1) better understand why severe COVID-19 patients develop stronger immune responses; 2) find potential cells for blockage or immune enhancement therapy or; 3) identify pathways that could be targeted pharmacologically.

**Analysis report from group 2:**
1. [project_report_covid19](single_cell/group_reports/project_report_covid19.html) ([.Rmd](single_cell/group_reports/project_report_covid19.Rmd))
2. [Presentation_covid19](single_cell/group_reports/project_presentation_covid19.pdf)


</details>

<br/>

# <img border="0" src="/single-cell_sib_scilifelab_2021/logos/long_read.png" width="40" height="40"> RNA velocity and trajectories
***

<details>
<summary>Click to expand!</summary>

  TO DO

</details>

<br/>

# <img border="0" src="/single-cell_sib_scilifelab_2021/logos/ribo_profiling.png" width="40" height="40"> Multi-omics integration
***

<details>
<summary>Click to expand!</summary>

  TO DO

</details>

<br/>

# <img border="0" src="/single-cell_sib_scilifelab_2021/logos/uv_crosslink_ip.png" width="40" height="40"> Deep learning
***

<details>
<summary>Click to expand!</summary>

  TO DO

</details>

<br/>

<br/>

### [Back to main](README.md)
