---
title: "Fetal brain cortex data (Trevino et al. 2020) - extract GluN trajectory"
output: html_notebook
---

```{r}
library(SingleCellExperiment)
library(scater)
library(ensembldb)
library(EnsDb.Hsapiens.v86)
```

Preparations

```{r}
hg38_genes <- genes(EnsDb.Hsapiens.v86)
path2data <- "~/brainchromatin/data_files/rds/"
```

## Load scRNA-seq data

```{r}
scrna_sce <- readRDS(paste0(path2data, "scRNA_SCE.RDS"))
colData(scrna_sce)
```

Save gene names

```{r}
gene_names <- rownames(scrna_sce)
gene_names[rownames(scrna_sce) %in% names(hg38_genes)] <- hg38_genes[rownames(scrna_sce) %in% names(hg38_genes)]$gene_name
rowData(scrna_sce)[["gene_name"]] <- gene_names
```

Subset to cells on differentiation trajectory of excitatory neurons (picking from the paper).

```{r}
traj_clusters <- c("c14", "c2", "c9", "c4", "c5", "c7", "c0", "c12", "c15", 'c21')
traj_cells <- colnames(scrna_sce)[scrna_sce$seurat_clusters %in% traj_clusters]
scrna_ext_traj_sce <- scrna_sce[,traj_cells]

plotUMAP(scrna_ext_traj_sce, colour_by="seurat_clusters", point_size=0.2, text_by="seurat_clusters")
```

## Load scATAC data
```{r}
scatac_sce <- readRDS(paste0(path2data, "scATAC_SCE.RDS"))
colData(scatac_sce)
## Reverse UMAP coords
reducedDim(scatac_sce, "UMAP")[,"UMAP2"] <- - reducedDim(scatac_sce, "UMAP")[,"UMAP2"]

```

```{r}
traj_clusters <- paste0("c",c(8,1,0,13,6,7,14,5,2))
traj_cells <- colnames(scatac_sce)[scatac_sce$Iterative.LSI.Clusters %in% traj_clusters]
scatac_ext_traj_sce <- scatac_sce[,traj_cells]

plotUMAP(scatac_ext_traj_sce, point_size=0.2, colour_by="Iterative.LSI.Clusters", text_by="Iterative.LSI.Clusters")
```

## Load multiome data

```{r}
multi_scrna_sce <- readRDS(paste0(path2data, "Multiome_RNA_SCE.RDS"))
multi_scatac_sce <- readRDS(paste0(path2data, "Multiome_ATAC_SCE.RDS"))

plotUMAP(multi_scrna_sce, point_size=0.2, colour_by="seurat_clusters", text_by="seurat_clusters")
```
Subset to Exc trajectory
```{r}
traj_clusters <- paste0("c",c(0,1,3,10,6))
traj_cells <- colnames(multi_scrna_sce)[multi_scrna_sce$seurat_clusters %in% traj_clusters]
multi_scrna_traj_sce <- multi_scrna_sce[,traj_cells]
multi_scatac_traj_sce <- multi_scatac_sce[,traj_cells]

plotUMAP(multi_scrna_traj_sce, point_size=0.2, colour_by="seurat_clusters", text_by="seurat_clusters")
```

Cleaning 

```{r}
## scRNA
keep_rowdata_colnames <- c('gene_name')
keep_coldata_colnames <- c("Cell.ID","Sample.ID","Age","Tissue.ID","Sample.Type",
                           "Assay", "Batch","seurat_clusters","RNA.Counts","RNA.Features",
                           "Percent.MT","Percent.Ribo","Cell.Barcode")
colData(scrna_ext_traj_sce) <- colData(scrna_ext_traj_sce)[keep_coldata_colnames]
rowData(scrna_ext_traj_sce) <- rowData(scrna_ext_traj_sce)[keep_rowdata_colnames]
reducedDims(scrna_ext_traj_sce) <- NULL

## scATAC
colData(scatac_ext_traj_sce) <- colData(scatac_ext_traj_sce)[!startsWith(colnames(colData(scatac_ext_traj_sce)), ".CR")]
rowData(scatac_ext_traj_sce) <- rowData(scatac_ext_traj_sce)[c("name", "classification")]
reducedDims(scatac_ext_traj_sce) <- NULL

## Multiome
keep_coldata_colnames_multiome <- c(0:10, 66)
colData(multi_scrna_traj_sce) <- colData(multi_scrna_traj_sce)[keep_coldata_colnames_multiome]
rowData(multi_scrna_traj_sce) <- rowData(multi_scrna_traj_sce)[c("gene_id", "gene_type", "gene_name")]
colData(multi_scatac_traj_sce) <- colData(multi_scatac_traj_sce)[keep_coldata_colnames_multiome]
rowData(multi_scatac_traj_sce) <- rowData(multi_scatac_traj_sce)[c("name")]
reducedDims(multi_scrna_traj_sce) <- NULL
```

## Save RDS objects

```{r}
saveRDS(scrna_ext_traj_sce, "~/single-cell_sib_scilifelab_2021/multiomics/processed_data/scRNA_ext_trajectory_SCE.RDS")
saveRDS(scatac_ext_traj_sce, "~/single-cell_sib_scilifelab_2021/multiomics/processed_data/scATAC_ext_trajectory_SCE.RDS")

saveRDS(multi_scrna_traj_sce, "~/single-cell_sib_scilifelab_2021/multiomics/processed_data/Multiome_RNA_ext_trajectory_SCE.RDS")
saveRDS(multi_scatac_traj_sce, "~/single-cell_sib_scilifelab_2021/multiomics/processed_data/Multiome_ATAC_ext_trajectory_SCE.RDS")

```

## Convert to anndata objects

Following the [Scanpy in R](https://theislab.github.io/scanpy-in-R/#setting-up) tutorial.

```{r}
# install.packages(renv)
renv::init()
renv::install("reticulate")
renv::use_python()

py_pkgs <- c(
    "scanpy",
    "anndata",
    "anndata2ri",
    "rpy2"
)

reticulate::py_install(py_pkgs)
```

```{r}
sc <- reticulate::import("scanpy")
convert_sce_2_anndata <- function(sce, assay, filename){
  if (assay=="RNA"){
    adata_sce <- sc$AnnData(
      X   = t(logcounts(sce)),
      obs = as.data.frame(colData(sce)),
      var = as.data.frame(rowData(sce)),
      layers = list(counts=t(counts(sce)))
      )
  } else {
    adata_sce <- sc$AnnData(
      X   = t(assay(sce)),
      obs = as.data.frame(colData(sce)),
      var = as.data.frame(rowRanges(sce))
      )
  }
  adata_sce$write_h5ad(filename)
}

outdir <- '/home/jovyan/single-cell_sib_scilifelab_2021/multiomics/processed_data/'
convert_sce_2_anndata(scrna_ext_traj_sce, "RNA", paste0(outdir, "scRNA_ext_trajectory_SCE.h5ad"))
convert_sce_2_anndata(scatac_ext_traj_sce, "ATAC", paste0(outdir, "scATAC_ext_trajectory_SCE.h5ad"))
convert_sce_2_anndata(multi_scrna_traj_sce, "RNA", paste0(outdir, "Multiome_RNA_ext_trajectory_SCE.h5ad"))
convert_sce_2_anndata(multi_scatac_traj_sce, "ATAC", paste0(outdir, "Multiome_ATAC_ext_trajectory_SCE.h5ad"))
```




