# <img border="0" src="../logos/omics_integration.png" width="40" height="40" style="vertical-align:middle;"> Omics Integration
***

<br/>

- __*Emma Dann*__, Wellcome Sanger Institute, Cambridge, ðŸ‡¬ðŸ‡§ United Kingdom
- __*Charlotte Soneson*__, Friedrich Miescher Institute, SIB/Basel, ðŸ‡¨ðŸ‡­ Switzerland

<br/>

### Pre-course

#### Software requirements 

We will be using an AWS cloud environment, where we have pre-installed useful software and downloaded the data.

Alternatively, you can set up your own working environment locally. To do so:

- clone the course code repository locally
```
git clone https://github.com/NBISweden/single-cell_sib_scilifelab_2021.git
```

- create a clone of our conda environment (if needed [install miniconda](https://conda.io/projects/conda/en/latest/user-guide/install/index.html) first):
```
cd single-cell_sib_scilifelab_2021/project_omics
conda env create --file multiomics-environment.yml
```

- download the preprocessed data from [GDrive](https://drive.google.com/drive/folders/1YjHfhxk2Z62pTEOTu27G-AgKqQawKEBT?usp=sharing) 

#### Knowledge requirements

We will have a warm-up session to start exploring our datasets before the project starts. 

You can start familiarizing yourself with certain tools we will be using, trying out the examples in vignettes:
* TBA
<!-- * muon:(https://muon-tutorials.readthedocs.io/en/latest/single-cell-rna-atac/index.html)
* MOFA2 -->

### Background

Corticogenesis is the dynamic process that results in the formation of the cerebral cortex, and is characterized by the generation of excitatory glutamatergic neurons from cortical progenitors, and the differentiation of astrocytes and oligodendrocytes. Dynamic changes in the activity of cis-regulatory DNA elements underlie the complex phenotypic transformations that occur during development.

Here we will be analyzing Human fetal brain cortex data from [Trevino et al. 2021](https://www.sciencedirect.com/science/article/abs/pii/S0092867421009429) ([source](https://github.com/GreenleafLab/brainchromatin)) ([OA preprint](https://www.biorxiv.org/content/10.1101/2020.12.29.424636v2.full)) to study the interplay between chromatin accessibility and gene expression in early corticogenesis.

Your main goal will be to identify non-coding genomic regions where chromatin accessibility is associated with expression of genes involved in excitatory neuron development.

- Group 1 will address the question with diagonal integration of the unmatched assay data (scRNA-seq data and scATAC-seq data from different cells)
- Group 2 will address the question with vertical integration of the multiome data (scRNA-seq and scATAC-seq for the same cells)

### Practical information

Each group will have a Jupyter Notebook showing initial exploration of the each omic dataset, with detailed instructions and helper code for the 



### Milestone 1:

1.1.

1.2.

1.3.

1.4.

<br/>

### Milestone 2:

2.1.

2.2.

2.3.

2.4.

2.5.

<br/>

### Milestone 3:

3.1.

3.2.

3.3.

<br/>

### Milestone 4:

4.1.

4.2.

4.3.

4.4.

4.5.

<br/>

### Milestone 5:

5.1.

5.2.

5.3.

<br/>

### Milestone 6:

6.1.

6.2.

6.3.

<br/>

### Milestone 7:

7.1.

7.2.

7.3.

<br/>

### Milestone 8:

8.1.1.

8.1.2.

8.1.3.

8.1.4.

8.1.5.
